core/zoe/contracts/publicAuction.js

```diff
@@ -20,75 +20,111 @@ export const makeContract = harden((zoe, terms) => {
   const ITEM_INDEX = 0;
   const BID_INDEX = 1;
 
-  const publicAuction = harden({
-    startAuction: async escrowReceipt => {
-      const {
-        offerHandle,
-        offerRules: { payoutRules },
-      } = await zoe.burnEscrowReceipt(escrowReceipt);
+  const makeBidderSeat = inviteHandle =>
+    harden({
+      bid: () => {
+        // Check that the item is still up for auction
+        const { inactive } = zoe.getOfferStatuses(
+          harden([creatorInviteHandle]),
+        );
+        if (inactive.length > 0) {
+          return rejectOffer(
+            zoe,
+            terms.assays,
+            inviteHandle,
+            'The item up for auction has been withdrawn or the auction has completed',
+          );
+        }
+
+        if (allBidHandles.length >= numBidsAllowed) {
+          return rejectOffer(
+            zoe,
+            terms.assays,
+            inviteHandle,
+            `No further bids allowed.`,
+          );
+        }
+
+        const [payoutRules] = zoe.getPayoutRuleMatrix(
+          harden([inviteHandle]),
+          terms.assays,
+        );
+
+        const ruleKinds = ['want', 'offer'];
+        if (!hasValidPayoutRules(ruleKinds, terms.assays, payoutRules)) {
+          return rejectOffer(zoe, terms.assays, inviteHandle);
+        }
 
-      const ruleKinds = ['offerExactly', 'wantAtLeast'];
+        if (
+          !isOverMinimumBid(
```

Comment:
"isOver" implies greater than. You should be checking that the bid is greater than or equal to the minimum bid. The name of this predicate should clearly reflect that.

In one of my previous sketches towards Zoe, I introduced a helper function `canTradeWith(payoutRules1, payoutRules2)` which just determined if a swap would preserve offer safety without deciding where the surplus should go. Enough contracts actually only to pairwise swap reallocations that it often proved useful, even though it doesn't cover all cases.

Reply:
Where do you check compatibility on the item? IOW, that the item the seller offers includes the item the bidder wants?

`canTradeWith` would cover both.

Reply:
Sure, that's a good idea. Do you know where `canTradeWith` can be found? I can reproduce it if not

Reply:
I could try to find it. But it was written for a system sufficiently different that you'll probably rewrite it from scratch anyway ;)